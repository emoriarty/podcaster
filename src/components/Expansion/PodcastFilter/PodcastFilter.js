import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { fetchPodcasts } from '../../../actions'
import { getCountry } from '../../../selectors/settings'
import { getMediaTypeUrl, getItemsLimit } from '../../../selectors/mediaTypes'
import { getPodcastsGenres } from '../../../selectors/genres'
import { getTranslations } from '../../../selectors/translations'
import { FilterPanel } from '..'
import SelectField from 'material-ui/SelectField'
import MenuItem from 'material-ui/MenuItem'
import Toggle from 'material-ui/Toggle'
import styles from './PodcastFilter.css'
import { equals } from 'ramda'

export class PodcastFilter extends Component {
  static propTypes = {
    genres: PropTypes.array,
    podcastsUrl: PropTypes.string,
    onChange: PropTypes.func,
    limits: PropTypes.array
  }

  constructor (props) {
    super(props)
    this.state = {
      genre: '',
      explicit: false,
      limit: 10
    }
    this.onChangeGenre = this.onChangeGenre.bind(this)
    this.onChangeExplicit = this.onChangeExplicit.bind(this)
    this.onChangeLimit = this.onChangeLimit.bind(this)
  }

  componentDidUpdate (prevProps, prevState) {
    const { country, fetchPodcasts, onChange, podcastsUrl } = this.props

    if (!equals(prevState, this.state)) {
      fetchPodcasts(podcastsUrl, country, this.state)
        .then(() =>
          onChange &&
            typeof onChange === 'function' &&
            onChange()
        )
    }
  }

  onChangeGenre (ev, index, value) {
    this.setState({ genre: value })
  }

  onChangeExplicit (ev, checked) {
    this.setState({ explicit: checked })
  }

  onChangeLimit (ev, index, value) {
    this.setState({ limit: value })
  }

  render () {
    const { genres, limits, translations } = this.props

    return (
      <FilterPanel label='Filter'>
        <form className={styles.podcastFilter}>
          <SelectField
            className={styles.selectField}
            floatingLabelFixed
            floatingLabelText={translations['Genre']}
            onChange={this.onChangeGenre}
            value={this.state.genre}>
            {
              genres.map((genre, index) =>
                <MenuItem
                  key={genre.id}
                  primaryText={genre.label}
                  value={genre.id} />
              )
            }
          </SelectField>
          <SelectField
            className={styles.selectField}
            floatingLabelFixed
            floatingLabelText={translations['Size']}
            onChange={this.onChangeLimit}
            value={this.state.limit}>
            {
              limits.map((limit, index) =>
                <MenuItem
                  key={index}
                  primaryText={limit}
                  value={limit} />
              )
            }
          </SelectField>
          <Toggle
            className={styles.toggle}
            label={translations['Explicit_Content']}
            onToggle={this.onChangeExplicit} />
        </form>
      </FilterPanel>
    )
  }
}

const mapStateToProps = state => {
  const country = getCountry(state)
  const genres = getPodcastsGenres(state)
  const translations = getTranslations(state)
  const podcastsUrl = getMediaTypeUrl(state, 'podcasts')
  const limits = getItemsLimit()

  return {
    country,
    genres,
    podcastsUrl,
    limits,
    translations
  }
}

const mapDispatchToProps = {
  fetchPodcasts
}

export default connect(mapStateToProps, mapDispatchToProps)(PodcastFilter)
