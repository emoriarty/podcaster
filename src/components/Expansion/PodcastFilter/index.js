export { default, default as PodcastFilter } from './PodcastFilter'
