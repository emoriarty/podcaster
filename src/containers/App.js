import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import config from '../../config/app'
import AppBar from 'material-ui/AppBar'
import FlatButton from 'material-ui/FlatButton'
import Avatar from 'material-ui/Avatar'
import {
  fetchCommonTranslations,
  fetchCountries,
  fetchMediaTypes,
  fetchMediaTypesTranslations,
  fetchPodcasts
} from '../actions'
import { pleaseWait } from 'please-wait'
import logo from '../../assets/images/logo-256.png'
import 'please-wait/src/please-wait.scss'

export class App extends Component {
  componentDidMount () {
    const {
      fetchCommonTranslations,
      fetchCountries,
      fetchMediaTypes,
      fetchMediaTypesTranslations,
      language
    } = this.props

    this.loadingScreen = pleaseWait({
      logo,
      backgroundColor: '#1976D2',
      loadingHtml: '<div class="spinner"><div class="double-bounce1"></div><div class="double-bounce2"></div></div>'
    })

    fetchCommonTranslations(language)
    fetchCountries()
    fetchMediaTypes()
    fetchMediaTypesTranslations(language)
  }

  componentWillReceiveProps (nextProps) {
    const { country, fetchPodcasts, podcastUrl } = this.props
    if (!nextProps.isLoading) {
      this.loadingScreen.finish()
    }
    console.log(nextProps.podcastUrl)
    if (nextProps.podcastUrl !== podcastUrl) {
      fetchPodcasts(nextProps.podcastUrl, country, 'limit=10/genre=1303/explicit=true')
    }
  }

  renderFlagButton () {
    const { countryFlag } = this.props

    return (
      <FlatButton>
        <Avatar
          style={{
            backgroundColor: 'transparent',
            marginTop: 3
          }}
          size={30}
          src={countryFlag}
        />
      </FlatButton>
    )
  }

  render () {
    const { isLoading } = this.props
    return (
      <div>
        {
          !isLoading &&
            <div>
              <AppBar
                showMenuIconButton={false}
                title='Podcaster'
                iconElementRight={this.renderFlagButton()}
              />
              <h2>Top</h2>
              <ul>
                <li>podcast 1</li>
              </ul>
            </div>
        }
      </div>
    )
  }
}

App.defaultProps = {
  isLoading: true
}

App.propTypes = {
  country: PropTypes.string,
  countryFlag: PropTypes.string,
  fetchCommonTranslations: PropTypes.func,
  fetchCountries: PropTypes.func,
  fetchMediaTypes: PropTypes.func,
  fetchMediaTypesTranslations: PropTypes.func,
  fetchPodcasts: PropTypes.func,
  language: PropTypes.string.isRequired,
  isLoading: PropTypes.bool.isRequired,
  podcastUrl: PropTypes.string
}

const mapStateToProps = (state) => {
  const countryCode = state.settings.country
  const countryFlag = countryCode in state.entities.countries
    ? state.entities.countries[countryCode].flag_icon
    : ''

  return {
    country: state.settings.country,
    countryFlag: `${config.itunes.provider.rss}${countryFlag}`,
    podcastUrl: '26' in state.entities.mediaTypes ? state.entities.mediaTypes['26'].feed_types.urlPrefix : '',
    isLoading: state.countries.isFetching ||
      state.mediaTypes.isFetching ||
      state.translations.common.isFetching ||
      state.translations.mediaTypes.isFetching,
    language: state.settings.language
  }
}

const mapDispatchToProps = (dispatch) =>
  bindActionCreators({
    fetchCommonTranslations,
    fetchCountries,
    fetchMediaTypes,
    fetchMediaTypesTranslations,
    fetchPodcasts
  }, dispatch)

export default connect(mapStateToProps, mapDispatchToProps)(App)
