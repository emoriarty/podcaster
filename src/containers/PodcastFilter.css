.podcast-filter {
  margin: 0 auto 32px auto;
  width: calc(100% - 32px);
}

.select-field {
  width: 100% !important;
}

.toggle {
  margin-top: 16px;
}
