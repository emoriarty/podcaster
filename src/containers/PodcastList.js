import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { fetchPodcasts } from '../actions'
import { getPodcasts } from '../selectors/podcasts'
import { getMediaTypeUrl } from '../selectors/mediaTypes'
import { getCountry } from '../selectors/settings'
import times from '../utils/times'
import { List } from 'material-ui/List'
import {
  TwoLineListItem,
  TwoLineLoadingListItem
} from '../components/ListItems'

class PodcastList extends Component {
  static propTypes = {
    country: PropTypes.string,
    podcasts: PropTypes.array,
    podcastsUrl: PropTypes.string
  }

  constructor (props) {
    super(props)

    this.renderItems = this.renderItems.bind(this)
    this.renderLoadingItems = this.renderLoadingItems.bind(this)
  }

  componentWillReceiveProps (nextProps) {
    // const { country, fetchPodcasts, podcastsUrl } = this.props
    //
    // if (nextProps.podcastsUrl !== podcastsUrl || nextProps.country !== country) {
    //   fetchPodcasts(nextProps.podcastsUrl, country, 'limit=50/explicit=true')
    // }
  }

  renderLoadingItems () {
    let items = []
    times(10)(
      index =>
        items.push(<TwoLineLoadingListItem key={index} />)
    )
    console.log(items)
    return items
  }

  renderItems () {
    const { podcasts } = this.props
    return podcasts.map(
      podcast =>
        <TwoLineListItem
          key={podcast['id.attributes.im:id']}
          primaryText={podcast['im:name.label']}
          secondaryText={podcast['im:artist.label']}
          avatar={podcast['im:image.0.label']}
        />
    )
  }

  render () {
    const { podcasts } = this.props

    return (
      <List>
        {
          podcasts.length > 0
            ? this.renderItems()
            : this.renderLoadingItems()
        }
      </List>
    )
  }
}

const mapStateToProps = state => {
  const country = getCountry(state)
  const podcasts = getPodcasts(state)
  const podcastsUrl = getMediaTypeUrl(state, 'podcasts')

  return {
    country,
    podcasts,
    podcastsUrl
  }
}

const mapDispatchToProps = {
  fetchPodcasts
}

export default connect(mapStateToProps, mapDispatchToProps)(PodcastList)
