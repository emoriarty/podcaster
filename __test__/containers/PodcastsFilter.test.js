import React from 'react'
import { PodcastFilter } from '../../src/containers/PodcastFilter'

const genres = [{
  id: '',
  label: 'all'
}, {
  id: '2',
  label: 'politics'
}, {
  id: '4',
  label: 'comedy'
}]

const limits = [10, 25, 50, 100]

describe('PodcastFilter container', () => {
  it('match snapshot', () => {
    const component = shallow(
      <PodcastFilter
        genres={genres}
        country='es'
        translations={{}}
        limits={limits} />)
    expect(component)
      .toMatchSnapshot()
  })
})
