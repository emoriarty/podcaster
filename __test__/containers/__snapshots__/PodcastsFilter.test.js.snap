// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PodcastFilter container match snapshot 1`] = `
<form>
  <SelectField
    autoWidth={false}
    disabled={false}
    floatingLabelFixed={true}
    fullWidth={false}
    multiple={false}
    onChange={[Function]}
    value=""
  >
    <MenuItem
      anchorOrigin={
        Object {
          "horizontal": "right",
          "vertical": "top",
        }
      }
      checked={false}
      desktop={false}
      disabled={false}
      focusState="none"
      insetChildren={false}
      primaryText="all"
      targetOrigin={
        Object {
          "horizontal": "left",
          "vertical": "top",
        }
      }
      value=""
    />
    <MenuItem
      anchorOrigin={
        Object {
          "horizontal": "right",
          "vertical": "top",
        }
      }
      checked={false}
      desktop={false}
      disabled={false}
      focusState="none"
      insetChildren={false}
      primaryText="politics"
      targetOrigin={
        Object {
          "horizontal": "left",
          "vertical": "top",
        }
      }
      value="2"
    />
    <MenuItem
      anchorOrigin={
        Object {
          "horizontal": "right",
          "vertical": "top",
        }
      }
      checked={false}
      desktop={false}
      disabled={false}
      focusState="none"
      insetChildren={false}
      primaryText="comedy"
      targetOrigin={
        Object {
          "horizontal": "left",
          "vertical": "top",
        }
      }
      value="4"
    />
  </SelectField>
  <SelectField
    autoWidth={false}
    disabled={false}
    floatingLabelFixed={true}
    fullWidth={false}
    multiple={false}
    onChange={[Function]}
    value={10}
  >
    <MenuItem
      anchorOrigin={
        Object {
          "horizontal": "right",
          "vertical": "top",
        }
      }
      checked={false}
      desktop={false}
      disabled={false}
      focusState="none"
      insetChildren={false}
      primaryText={10}
      targetOrigin={
        Object {
          "horizontal": "left",
          "vertical": "top",
        }
      }
      value={10}
    />
    <MenuItem
      anchorOrigin={
        Object {
          "horizontal": "right",
          "vertical": "top",
        }
      }
      checked={false}
      desktop={false}
      disabled={false}
      focusState="none"
      insetChildren={false}
      primaryText={25}
      targetOrigin={
        Object {
          "horizontal": "left",
          "vertical": "top",
        }
      }
      value={25}
    />
    <MenuItem
      anchorOrigin={
        Object {
          "horizontal": "right",
          "vertical": "top",
        }
      }
      checked={false}
      desktop={false}
      disabled={false}
      focusState="none"
      insetChildren={false}
      primaryText={50}
      targetOrigin={
        Object {
          "horizontal": "left",
          "vertical": "top",
        }
      }
      value={50}
    />
    <MenuItem
      anchorOrigin={
        Object {
          "horizontal": "right",
          "vertical": "top",
        }
      }
      checked={false}
      desktop={false}
      disabled={false}
      focusState="none"
      insetChildren={false}
      primaryText={100}
      targetOrigin={
        Object {
          "horizontal": "left",
          "vertical": "top",
        }
      }
      value={100}
    />
  </SelectField>
  <Toggle
    defaultToggled={false}
    disabled={false}
    labelPosition="left"
    onToggle={[Function]}
  />
</form>
`;
